pkg/apis/apps/v1alpha1/apimanager_types_test.go

```diff
@@ -0,0 +1,113 @@
+	"github.com/3scale/3scale-operator/version"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestSetDefaults(t *testing.T) {
+	tmpDefaultAppLabel := defaultAppLabel
+	tmpDefaultTenantName := defaultTenantName
+	tmpDefaultImageStreamTagImportInsecure := defaultImageStreamImportInsecure
+	tmpDefaultResourceRequirementsEnabled := defaultResourceRequirementsEnabled
+	tmpDefaultApicastManagementAPI := defaultApicastManagementAPI
+	tmpDefaultApicastOpenSSLVerify := defaultApicastOpenSSLVerify
+	tmpDefaultApicastResponseCodes := defaultApicastResponseCodes
+	tmpDefaultApicastRegistryURL := defaultApicastRegistryURL
+
+	var tmpDefaultReplicas int64 = 1
+
+	inputAPIManager := APIManager{
+		Spec: APIManagerSpec{
+			APIManagerCommonSpec: APIManagerCommonSpec{
+				WildcardDomain: "test.3scale.com",
+			},
+		},
+	}
+	expectedAPIManager := APIManager{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				OperatorVersionAnnotation:   version.Version,
+				ThreescaleVersionAnnotation: product.ThreescaleRelease,
+			},
+		},
+		Spec: APIManagerSpec{
+			APIManagerCommonSpec: APIManagerCommonSpec{
+				WildcardDomain:               "test.3scale.com",
+				AppLabel:                     &tmpDefaultAppLabel,
+				TenantName:                   &tmpDefaultTenantName,
+				ImageStreamTagImportInsecure: &tmpDefaultImageStreamTagImportInsecure,
+				ResourceRequirementsEnabled:  &tmpDefaultResourceRequirementsEnabled,
+			},
+			Apicast: &ApicastSpec{
+				IncludeResponseCodes: &tmpDefaultApicastResponseCodes,
+				ApicastManagementAPI: &tmpDefaultApicastManagementAPI,
+				OpenSSLVerify:        &tmpDefaultApicastOpenSSLVerify,
+				RegistryURL:          &tmpDefaultApicastRegistryURL,
+				ProductionSpec: &ApicastProductionSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+				StagingSpec: &ApicastStagingSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+			},
+			Backend: &BackendSpec{
+				ListenerSpec: &BackendListenerSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+				WorkerSpec: &BackendWorkerSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+				CronSpec: &BackendCronSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+			},
+			System: &SystemSpec{
+				AppSpec: &SystemAppSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+				DatabaseSpec: &SystemDatabaseSpec{
+					MySQL: &SystemMySQLSpec{},
+				},
+				FileStorageSpec: &SystemFileStorageSpec{
+					PVC: &SystemPVCSpec{},
+				},
+				SidekiqSpec: &SystemSidekiqSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+			},
+			Zync: &ZyncSpec{
+				AppSpec: &ZyncAppSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+				QueSpec: &ZyncQueSpec{
+					Replicas: &tmpDefaultReplicas,
+				},
+			},
+		},
+	}
+
+	t.Run("Basic defaults are correct", testBasicAPIManagerDefaults(&inputAPIManager, &expectedAPIManager))
+}
+
+func testBasicAPIManagerDefaults(input, expected *APIManager) func(t *testing.T) {
+	return func(t *testing.T) {
+		tmpInput := input.DeepCopy()
```

Comment:
Here we need to copy the input because the SetDefault methods modifies it

Reply:
If you define the input inside, it would not be shared among other subtests.

I would do the `DeepCopy` outside subtest. This way high level test ensures all subtests get expected input data structure and not modified by some subtest

Reply:
makes sense :+1: I'll change it

Reply:
Changed